### DISCUSSION TITLE
Is there a way to retry an API operation? C# Azure

### DISCUSSION THREAD

**Discussion by karpikpl** [pulumi/pulumi#10692] [Pulumi in Practice]
I have an operation that sometimes works, sometimes fails in Azure and I have to re-run pulumi-up.

```csharp
        // sleep 10s because Azure... List Host Keys method sometimes fails with HTTP 400, re-running stack fixes it.
        if (!Deployment.Instance.IsDryRun)
        {
            Log.Info("Waiting 10s before calling function to get host keys");
            System.Threading.Thread.Sleep(10000);
        }

        var keys = Output.CreateSecret(ListWebAppHostKeys.Invoke(new ListWebAppHostKeysInvokeArgs
        {
            Name = functionApp.Name,
            ResourceGroupName = args.ResourceGroupName
        }, new InvokeOptions { Parent = functionApp }));
```

I tried wrapping `ListWebAppHostKeys` into some try/catch block but that doesn't seem to be possible.

Ideally I would like to use **Polly** and deal with custom retries.
I can try making that **List Host Keys** with Azure SDK, but I'm not sure if that's going to work.

**Comment by karpikpl** [comment]
I've ended up implementing that call outside of Pulumi. Hope it helps someone with similar issues.
```csharp
    public Output<string> GetHostKeys(Output<string> rgName, Output<string> functionName)
    {
        return Output.Tuple(rgName, functionName).Apply(async t =>
        {
            var (resourceGroupName, siteName) = t;

            var config = await GetClientConfig.InvokeAsync();

            var result = await ApiClient
                .WithRetry(count: 4, seconds: 5)
                // Azure returns 400 BadRequest when Function App is not ready
                .WithCustomRetryPolicy(HttpPolicyExtensions.HandleTransientHttpError().OrResult(http => http.StatusCode == System.Net.HttpStatusCode.BadRequest || http.StatusCode == System.Net.HttpStatusCode.NotFound))
                .PostAsync<HostKeys>($"https://management.azure.com/subscriptions/{config.SubscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{siteName}/host/default/listkeys?api-version=2022-03-01");

            return result.Value.FunctionKeys.Key;
        });
    }
```
I'm using open source library for making that call (that's also where my infra project is)
https://github.com/Avanade/CoreEx/tree/main/samples/My.Hr/My.Hr.Infra

I'm not closing the discussion because I would to get some feedback, if going custom with retries is the only solution.

**Comment by PabloZaiden** [comment]
We're facing the same issue. It seems that the problem here is that the webapp is not ready soon enough to retrieve the keys. 
Having either a way to specify retries or being able to add an explicit delay would probably solve it.

**Reply by karpikpl** [reply]
I tried with explicit delay (Thread.Sleep) but it's hard to predict when app is going to be ready...

Ability to define a retry policy for API calls (not resource creation) would be ideal.

**Reply by PabloZaiden** [reply]
For sure. Either an explicit retry policy or an internal validation that the resource is ready would be ideal.
